apps/dotcom/src/utils/assetHandler.ts

```diff
@@ -0,0 +1,33 @@
+import { AssetContextProps, TLAsset } from 'tldraw'
+import { ASSET_UPLOADER_URL } from './config'
+
+export async function resolveAsset(asset: TLAsset | null | undefined, context: AssetContextProps) {
+	if (!asset || !asset.props.src) return ''
+
+	// We don't deal with videos at the moment.
+	if (asset.type === 'video') return asset.props.src
+
+	// Assert it's an image to make TS happy.
+	if (asset.type !== 'image') return ''
+
+	// Don't try to transform data: URLs, yikes.
+	if (!asset.props.src.startsWith('http:') && !asset.props.src.startsWith('https:'))
+		return asset.props.src
+
+	// N.B. navigator.connection is only available in certain browsers (mainly Blink-based browsers)
+	// 4g is as high the 'effectiveType' goes and we can pick a lower effective image quality for slower connections.
+	const networkCompensation =
+		!context.networkEffectiveType || context.networkEffectiveType === '4g' ? 1 : 0.5
+
+	// We only look at the zoom level to the nearest 0.25
+	const zoomStepFunction = (zoom: number) => Math.floor(zoom * 4) / 4
+	const steppedZoom = Math.max(0.25, zoomStepFunction(context.zoom))
+
+	const width = Math.ceil(asset.props.w * steppedZoom * networkCompensation)
```

Comment:
no no, i think you misread: the check is for `!context.networkEffectiveType` - non-Blink browsers always get a multiplier of `1`.

so the _only_ case where that cames into play is: Blink browsers + 2g/3g

Reply:
oh I see! that makes sense. I still think it could be worth using this with the quality setting instead though (the default is `85`, we set it to `92` for some reason)

Reply:
yeah, you may be right on using `quality` - i mean, truth be told, i'm stabbing in the dark a bit at the moment.
what really needs to happen first is I need to test this on Cloudflare (after Steve enables it).

i think i was not changing `quality` to keep the surface area limited at the moment. but yeah, once i can actually test this work on Cloudflare then i'll see if `quality` is a better parameter to tweak vs. width as far as output. i'll circle back to this one!

side note: i picked `92` to keep parity with this: https://github.com/tldraw/tldraw/blob/main/packages/tldraw/src/lib/defaultExternalContentHandlers.ts#L88

Reply:
now that i can play around with this it looks `quality` doesn't behave quite how we'd expect; for that reason, we should stick with tweaking `width` instead.

example image:
- 1000px width, quality 92, 604kb: https://assets.tldraw.xyz/cdn-cgi/image/width=1000,dpr=2,fit=scale-down,quality=92/https://assets.tldraw.xyz/uploads/ypeqp-SgVcRYsA18ccwMF-Screen-Shot-2021-08-16-at-8.48.36-AM.png
- 500px width, quality 92, 176kb: https://assets.tldraw.xyz/cdn-cgi/image/width=500,dpr=2,fit=scale-down,quality=92/https://assets.tldraw.xyz/uploads/ypeqp-SgVcRYsA18ccwMF-Screen-Shot-2021-08-16-at-8.48.36-AM.png

---

- 1000px width, quality 85, 604kb (same as quality 92!): https://assets.tldraw.xyz/cdn-cgi/image/width=1000,dpr=2,fit=scale-down,quality=85/https://assets.tldraw.xyz/uploads/ypeqp-SgVcRYsA18ccwMF-Screen-Shot-2021-08-16-at-8.48.36-AM.png
- 1000px width, quality 60, 369kb: https://assets.tldraw.xyz/cdn-cgi/image/width=1000,dpr=2,fit=scale-down,quality=60/https://assets.tldraw.xyz/uploads/ypeqp-SgVcRYsA18ccwMF-Screen-Shot-2021-08-16-at-8.48.36-AM.png
- 1000px width, quality 30, 229kb: https://assets.tldraw.xyz/cdn-cgi/image/width=1000,dpr=2,fit=scale-down,quality=30/https://assets.tldraw.xyz/uploads/ypeqp-SgVcRYsA18ccwMF-Screen-Shot-2021-08-16-at-8.48.36-AM.png